libyul/Object.cpp

```diff
@@ -144,33 +145,61 @@ set<YulString> Object::qualifiedDataNames() const
 	return qualifiedNames;
 }
 
-vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+void Object::traverseObjectTree(Object const* _object, YulString _qualifiedName, std::function<bool(Object const*)> _visitor)
 {
-	yulAssert(_qualifiedName != name, "");
-	yulAssert(subIndexByName.count(name) == 0, "");
-
-	if (boost::algorithm::starts_with(_qualifiedName.str(), name.str() + "."))
-		_qualifiedName = YulString{_qualifiedName.str().substr(name.str().length() + 1)};
 	yulAssert(!_qualifiedName.empty(), "");
 
 	vector<string> subObjectPathComponents;
 	boost::algorithm::split(subObjectPathComponents, _qualifiedName.str(), boost::is_any_of("."));
 
-	vector<size_t> path;
-	Object const* object = this;
 	for (string const& currentSubObjectName: subObjectPathComponents)
 	{
 		yulAssert(!currentSubObjectName.empty(), "");
-		auto subIndexIt = object->subIndexByName.find(YulString{currentSubObjectName});
-		yulAssert(
-			subIndexIt != object->subIndexByName.end(),
-			"Assembly object <" + _qualifiedName.str() + "> not found or does not contain code."
+
+		auto subObjectIt = ranges::find_if(
+			_object->subObjects,
+			[&currentSubObjectName](auto const& _subObject) { return _subObject->name.str() == currentSubObjectName; }
 		);
```

Comment:
This replaces search in a hash map with linear search through a vector. The old version was both more efficient (for large objects) and more concise.

The `if` below should also be an assert.

Reply:
Done. I added the `operator::at` instead of `operator[]` since to keep the same semantics of the later, we would need to add a new element if it does not exists.